types/src/transaction.rs

```diff
@@ -0,0 +1,785 @@
+#[cfg(feature = "json-schema")]
+use schemars::JsonSchema;
+#[cfg(any(feature = "std", test))]
+use serde::{Deserialize, Serialize};
+use tracing::debug;
+
+#[cfg(any(all(feature = "std", feature = "testing"), test))]
+use crate::testing::TestRng;
+use crate::{
+    bytesrepr::{self, FromBytes, ToBytes},
+    crypto, Digest, DisplayIter, PublicKey, SecretKey, TimeDiff, Timestamp,
+};
+#[cfg(any(feature = "std", test))]
+use crate::{AccountAndSecretKey, TransactionConfig};
+pub use auction_transaction::AuctionTransaction;
+pub use direct_call::DirectCall;
+pub use error::{
+    DecodeFromJsonError as TransactionDecodeFromJsonError, Error as TransactionError,
+    ExcessiveSizeError as TransactionExcessiveSizeError, TransactionConfigFailure,
+};
+pub use native_transaction::NativeTransaction;
+pub use pricing_mode::PricingMode;
+#[cfg(any(all(feature = "std", feature = "testing"), test))]
+pub use test_transaction_builder::TestTransactionBuilder;
+pub use transaction_approval::TransactionApproval;
+#[cfg(any(feature = "std", test))]
+pub use transaction_builder::{TransactionBuilder, TransactionBuilderError};
+pub use transaction_hash::TransactionHash;
+pub use transaction_header::TransactionHeader;
+pub use transaction_kind::TransactionKind;
+pub use userland_transaction::UserlandTransaction;
+
+#[cfg(feature = "json-schema")]
+static TRANSACTION: Lazy<Transaction> = Lazy::new(|| {
+    let secret_key = SecretKey::example();
+    let args = crate::runtime_args! { "amount" => 1000 };
+    Transaction::build(
+        *Timestamp::example(),
+        TimeDiff::from_seconds(3_600),
+        PricingMode::GasPriceMultiplier(1),
+        String::from("casper-example"),
+        TransactionKind::from(NativeTransaction::new_mint_transfer(args)),
+        AccountAndSecretKey::SecretKey(secret_key),
+    )
+});
+
+/// A signed smart contract.
+///
+/// To construct a new `Transaction`, use a [`TransactionBuilder`].
+#[derive(Clone, Eq, Debug)]
+#[cfg_attr(
+    any(feature = "std", test),
+    derive(Serialize, Deserialize),
+    serde(deny_unknown_fields)
+)]
+#[cfg_attr(feature = "datasize", derive(DataSize))]
+#[cfg_attr(
+    feature = "json-schema",
+    derive(JsonSchema),
+    schemars(description = "A signed smart contract.")
+)]
+pub struct Transaction {
+    hash: TransactionHash,
+    header: TransactionHeader,
+    body: TransactionKind,
+    approvals: BTreeSet<TransactionApproval>,
+    #[cfg_attr(any(all(feature = "std", feature = "once_cell"), test), serde(skip))]
+    #[cfg_attr(
+        all(any(feature = "once_cell", test), feature = "datasize"),
+        data_size(skip)
+    )]
+    #[cfg(any(feature = "once_cell", test))]
+    is_verified: OnceCell<Result<(), TransactionConfigFailure>>,
+}
+
+impl Transaction {
+    /// Called by the `TransactionBuilder` to construct a new `Transaction`.
+    #[cfg(any(feature = "std", test))]
+    fn build(
+        timestamp: Timestamp,
+        ttl: TimeDiff,
+        pricing_mode: PricingMode,
```

Comment:
I know that this is where this discussion has tended, but this does only make sense if there are multiple lanes potentially accommodating the same TransactionKind with different pricing. I suppose one could have, just for UserlandTransaction type, the "FIFO" and "competitive" lanes at the least, so this is reasonable, but we should consider far in advance what configuration of lanes we actually want, besides "transfers, installers, everything else and _maybe_ reserved_".